File: src/lib/github.ts
import type { GitHubRepo, HelmChart, ChartDependency, ChartFile } from "./types";
import yaml from "js-yaml";

const GITHUB_API = "https://api.github.com";

const cache = new Map<string, { data: any; ts: number }>();
const CACHE_TTL = 5 * 60 * 1000;

// Module-level token — set via setAuthToken() from the auth context
let _authToken: string | null = null;

export function setAuthToken(token: string | null) {
  _authToken = token;
}

async function ghFetch(url: string, token?: string): Promise<Response> {
  const effectiveToken = token ?? _authToken;
  const headers: Record<string, string> = { Accept: "application/vnd.github+json" };
  if (effectiveToken) {
    headers.Authorization = `Bearer ${effectiveToken}`;
  }
  const res = await fetch(url, { headers });
  if (res.status === 403 || res.status === 429) {
    throw new Error("GitHub API rate limit reached. Please wait a minute and try again.");
  }
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.message || `GitHub API error (${res.status})`);
  }
  return res;
}

async function ghFetchJson<T = any>(url: string, token?: string): Promise<T> {
  const cacheKey = `${url}:${token ?? _authToken ?? "anon"}`;
  const cached = cache.get(cacheKey);
  if (cached && Date.now() - cached.ts < CACHE_TTL) {
    return cached.data as T;
  }
  const res = await ghFetch(url, token);
  const data = await res.json();
  cache.set(cacheKey, { data, ts: Date.now() });
  return data as T;
}

const OWNER_REPO_RE = /^([a-zA-Z0-9_.-]+)\/([a-zA-Z0-9_.-]+?)(?:\/.*)?$/;

export async function searchRepos(query: string): Promise<GitHubRepo[]> {
  if (!query.trim()) return [];

  // If query looks like "org/repo" or "org/repo/subdir", try exact lookup first
  const match = query.trim().match(OWNER_REPO_RE);
  if (match) {
    try {
      const repo = await ghFetchJson<GitHubRepo>(`${GITHUB_API}/repos/${match[1]}/${match[2]}`);
      return [repo];
    } catch {
      // Exact match failed — fall through to search
    }
  }

  const data = await ghFetchJson<{ items: GitHubRepo[] }>(
    `${GITHUB_API}/search/repositories?q=${encodeURIComponent(query)}&sort=stars&per_page=8`
  );
  return data.items;
}

export function parseRepoUrl(url: string): { owner: string; repo: string; subpath?: string } | null {
  const match = url.match(/github\.com\/([^/]+)\/([^/]+?)(?:\.git)?(?:\/tree\/[^/]+\/(.+?))?(?:\/)?$/);
  if (match) {
    return {
      owner: match[1],
      repo: match[2],
      subpath: match[3] || undefined,
    };
  }
  return null;
}

export async function getRepoByFullName(fullName: string): Promise<GitHubRepo> {
  return ghFetchJson<GitHubRepo>(`${GITHUB_API}/repos/${fullName}`);
}

async function searchTree(owner: string, repo: string, branch: string, subpath?: string): Promise<string[]> {
  if (subpath) {
    return searchContentsRecursive(owner, repo, subpath);
  }
  const data = await ghFetchJson<{ truncated?: boolean; tree: { type: string; path: string }[] }>(
    `${GITHUB_API}/repos/${owner}/${repo}/git/trees/${branch}?recursive=1`
  );
  if (data.truncated) {
    throw new Error("Repository tree is too large. Try pasting a URL that points to a specific chart directory.");
  }
  return data.tree
    .filter((item) => item.type === "blob")
    .map((item) => item.path);
}

async function searchContentsRecursive(owner: string, repo: string, path: string): Promise<string[]> {
  const data = await ghFetchJson<any>(`${GITHUB_API}/repos/${owner}/${repo}/contents/${path}`);
  if (!Array.isArray(data)) return [data.path];
  const files: string[] = [];
  for (const item of data) {
    if (item.type === "file") {
      files.push(item.path);
    } else if (item.type === "dir") {
      const subData = await ghFetchJson<any>(`${GITHUB_API}/repos/${owner}/${repo}/contents/${item.path}`);
      if (Array.isArray(subData)) {
        for (const sub of subData) {
          if (sub.type === "file") files.push(sub.path);
        }
      }
    }
  }
  return files;
}

export async function findHelmCharts(owner: string, repo: string, subpath?: string): Promise<HelmChart[]> {
  const repoData = await ghFetchJson<{ default_branch: string }>(`${GITHUB_API}/repos/${owner}/${repo}`);
  const branch = repoData.default_branch;

  const files = await searchTree(owner, repo, branch, subpath);
  const chartYamls = files.filter((f) => f.endsWith("Chart.yaml"));

  const charts: HelmChart[] = [];
  for (const chartPath of chartYamls) {
    try {
      const content = await getFileContent(owner, repo, chartPath);
      const parsed = yaml.load(content) as Record<string, any>;
      const deps: ChartDependency[] = (parsed.dependencies || [])
        .filter((d: any) => typeof d === "object")
        .map((d: any) => ({
          name: d.name || "",
          version: String(d.version || ""),
          repository: d.repository || "",
          condition: d.condition || "",
        }));
      charts.push({
        name: parsed.name || chartPath,
        version: parsed.version || "unknown",
        description: parsed.description || "",
        path: chartPath.replace("/Chart.yaml", ""),
        dependencies: deps,
      });
    } catch {
      charts.push({
        name: chartPath,
        version: "unknown",
        description: "",
        path: chartPath.replace("/Chart.yaml", ""),
      });
    }
  }
  return charts;
}

export async function getFileContent(owner: string, repo: string, path: string): Promise<string> {
  const data = await ghFetchJson<{ content: string }>(`${GITHUB_API}/repos/${owner}/${repo}/contents/${path}`);
  return atob(data.content);
}

const CHART_FILE_EXTENSIONS = new Set([
  ".yaml", ".yml", ".json", ".toml", ".tpl", ".txt", ".md", ".helmignore", ".lock",
]);

function hasAllowedExtension(filePath: string): boolean {
  const lower = filePath.toLowerCase();
  if (lower.endsWith(".helmignore") || lower.endsWith("chart.lock")) return true;
  const dotIdx = lower.lastIndexOf(".");
  if (dotIdx === -1) return false;
  return CHART_FILE_EXTENSIONS.has(lower.slice(dotIdx));
}

async function listFilesRecursive(owner: string, repo: string, path: string): Promise<string[]> {
  const data = await ghFetchJson<any>(`${GITHUB_API}/repos/${owner}/${repo}/contents/${path}`);
  if (!Array.isArray(data)) return [data.path];
  const files: string[] = [];
  const subdirs: string[] = [];
  for (const item of data) {
    if (item.type === "file") {
      files.push(item.path);
    } else if (item.type === "dir") {
      subdirs.push(item.path);
    }
  }
  const subResults = await Promise.all(
    subdirs.map((dir) => listFilesRecursive(owner, repo, dir))
  );
  for (const sub of subResults) {
    files.push(...sub);
  }
  return files;
}

export async function fetchChartFiles(
  owner: string,
  repo: string,
  chartPath: string,
): Promise<ChartFile[]> {
  const allPaths = await listFilesRecursive(owner, repo, chartPath);
  const prefix = chartPath.endsWith("/") ? chartPath : chartPath + "/";
  const filteredPaths = allPaths.filter((p) => {
    const rel = p.startsWith(prefix) ? p.slice(prefix.length) : p;
    if (rel === "values.yaml") return false;
    return hasAllowedExtension(p);
  });

  const results: ChartFile[] = [];
  const CONCURRENCY = 5;
  for (let i = 0; i < filteredPaths.length; i += CONCURRENCY) {
    const batch = filteredPaths.slice(i, i + CONCURRENCY);
    const settled = await Promise.allSettled(
      batch.map(async (filePath) => {
        const content = await getFileContent(owner, repo, filePath);
        const relativePath = filePath.startsWith(prefix) ? filePath.slice(prefix.length) : filePath;
        return { relativePath, content };
      })
    );
    for (const result of settled) {
      if (result.status === "fulfilled") {
        results.push(result.value);
      }
    }
  }
  return results;
}

export async function getUserRepos(token: string): Promise<GitHubRepo[]> {
  return ghFetchJson<GitHubRepo[]>(
    `${GITHUB_API}/user/repos?sort=updated&per_page=20&type=owner`,
    token,
  );
}

/**
 * Check if a repo likely contains a Helm chart by searching for Chart.yaml.
 * Uses the GitHub code search API with a per-repo scope.
 * Returns true if Chart.yaml is found, false otherwise.
 */
export async function repoHasHelmChart(owner: string, repo: string, token: string): Promise<boolean> {
  try {
    const data = await ghFetchJson<{ total_count: number }>(
      `${GITHUB_API}/search/code?q=filename:Chart.yaml+repo:${owner}/${repo}&per_page=1`,
      token,
    );
    return data.total_count > 0;
  } catch {
    return false;
  }
}

File: app/github.py
import base64
import io
import re
import tarfile
from dataclasses import dataclass, field

import httpx
import yaml

from app.config import GITHUB_TOKEN

_GITHUB_HEADERS: dict[str, str] = {"Accept": "application/vnd.github.v3+json"}


def _gh_headers() -> dict[str, str]:
    headers = dict(_GITHUB_HEADERS)
    if GITHUB_TOKEN:
        headers["Authorization"] = f"Bearer {GITHUB_TOKEN}"
    return headers


@dataclass
class ChartInfo:
    org: str
    repo: str
    branch: str
    directory: str
    chart_yaml: str
    values_yaml: str | None
    chart_name: str = ""
    chart_version: str = ""
    chart_description: str = ""
    app_version: str = ""
    dependencies: list[dict] = field(default_factory=list)
    helm_repo_url: str = ""  # non-empty when chart comes from a Helm registry (not GitHub)


def parse_github_url(url: str) -> tuple[str, str, str, str]:
    """Parse a GitHub URL into (org, repo, branch, directory).

    Supports formats:
      https://github.com/org/repo/tree/branch/path/to/chart
      https://github.com/org/repo (defaults to main, root dir)
    """
    url = url.strip().rstrip("/")

    pattern = r"https?://github\.com/([^/]+)/([^/]+)(?:/tree/([^/]+)(?:/(.+))?)?"
    match = re.match(pattern, url)
    if not match:
        raise ValueError(
            f"Invalid GitHub URL: {url}. "
            "Expected format: https://github.com/org/repo/tree/branch/path"
        )

    org = match.group(1)
    repo = match.group(2)
    branch = match.group(3) or "main"
    directory = match.group(4) or "."

    return org, repo, branch, directory


def is_github_url(text: str) -> bool:
    """Check whether text looks like a GitHub URL (vs a search query)."""
    text = text.strip()
    return text.startswith("http://github.com") or text.startswith("https://github.com")


def _parse_chart_yaml(content: str) -> dict:
    """Extract name, version, description, appVersion, and dependencies from Chart.yaml."""
    try:
        data = yaml.safe_load(content) or {}
    except yaml.YAMLError:
        data = {}

    result: dict = {
        "name": data.get("name", ""),
        "version": str(data.get("version", "")),
        "description": data.get("description", ""),
        "appVersion": str(data.get("appVersion", "")),
        "dependencies": [],
    }

    for dep in data.get("dependencies", []):
        if isinstance(dep, dict):
            result["dependencies"].append({
                "name": dep.get("name", ""),
                "version": str(dep.get("version", "")),
                "repository": dep.get("repository", ""),
                "condition": dep.get("condition", ""),
            })

    return result


async def search_repos(query: str) -> list[dict]:
    """Search GitHub for repos containing Helm charts.

    Uses code search with filename:Chart.yaml so every result is guaranteed
    to contain at least one Helm chart. Falls back to repo search with a
    topic:helm-chart qualifier if code search returns nothing.
    """
    async with httpx.AsyncClient(timeout=15.0) as client:
        # Primary: code search for Chart.yaml files matching the query
        # (requires authentication — falls back to repo search if no token)
        data = {"items": []}
        if GITHUB_TOKEN:
            resp = await client.get(
                "https://api.github.com/search/code",
                params={"q": f"{query} filename:Chart.yaml", "per_page": 20},
                headers=_gh_headers(),
            )
            if resp.status_code == 403:
                raise RuntimeError("GitHub API rate limit exceeded. Try again shortly.")
            if resp.status_code == 200:
                data = resp.json()

        # Deduplicate by repo (code search returns one result per file)
        seen: set[str] = set()
        results: list[dict] = []
        for item in data.get("items", []):
            repo = item.get("repository", {})
            name = repo.get("full_name", "")
            if name in seen:
                continue
            seen.add(name)
            results.append({
                "full_name": name,
                "description": repo.get("description") or "",
                "stars": repo.get("stargazers_count", 0),
                "avatar_url": repo.get("owner", {}).get("avatar_url", ""),
                "default_branch": repo.get("default_branch", "main"),
            })
            if len(results) >= 8:
                break

        if results:
            return results

        # Fallback: repo search with helm-chart topic
        resp = await client.get(
            "https://api.github.com/search/repositories",
            params={"q": f"{query} topic:helm-chart", "sort": "stars", "per_page": 8},
            headers=_gh_headers(),
        )
        if resp.status_code == 403:
            raise RuntimeError("GitHub API rate limit exceeded. Try again shortly.")
        resp.raise_for_status()
        data = resp.json()

    results = []
    for item in data.get("items", []):
        results.append({
            "full_name": item["full_name"],
            "description": item.get("description") or "",
            "stars": item.get("stargazers_count", 0),
            "avatar_url": item.get("owner", {}).get("avatar_url", ""),
            "default_branch": item.get("default_branch", "main"),
        })
    return results


async def discover_charts(org: str, repo: str, branch: str) -> list[dict]:
    """Use GitHub Trees API to find all Chart.yaml files and parse them.

    Returns list of {name, version, description, path} per chart. Limit 25.
    """
    async with httpx.AsyncClient(timeout=30.0) as client:
        # Get the full tree recursively
        resp = await client.get(
            f"https://api.github.com/repos/{org}/{repo}/git/trees/{branch}",
            params={"recursive": "1"},
            headers=_gh_headers(),
        )
        if resp.status_code == 403:
            raise RuntimeError("GitHub API rate limit exceeded. Try again shortly.")
        resp.raise_for_status()
        tree = resp.json()

        # Find all Chart.yaml files
        chart_paths = []
        for item in tree.get("tree", []):
            if item.get("type") == "blob" and item["path"].endswith("Chart.yaml"):
                chart_paths.append(item["path"])
            if len(chart_paths) >= 25:
                break

        # Fetch and parse each Chart.yaml
        charts = []
        for path in chart_paths:
            content = await fetch_file(client, org, repo, path, branch)
            if content is None:
                continue
            info = _parse_chart_yaml(content)
            # Derive the chart directory (parent of Chart.yaml)
            directory = "/".join(path.split("/")[:-1]) or "."
            charts.append({
                "name": info.get("name", path.rsplit("/", 1)[0] if "/" in path else repo),
                "version": info.get("version", ""),
                "description": info.get("description", ""),
                "path": directory,
                "dependencies": info.get("dependencies", []),
            })

    return charts


async def search_artifacthub(query: str) -> list[dict]:
    """Search ArtifactHub for Helm charts. Returns list of chart info dicts."""
    async with httpx.AsyncClient(timeout=15.0) as client:
        resp = await client.get(
            "https://artifacthub.io/api/v1/packages/search",
            params={"ts_query_web": query, "kind": 0, "limit": 8},
        )
        resp.raise_for_status()
        data = resp.json()

    results = []
    for pkg in data.get("packages", []):
        repo_info = pkg.get("repository", {})
        results.append({
            "name": pkg.get("name", ""),
            "description": pkg.get("description", ""),
            "version": pkg.get("version", ""),
            "stars": pkg.get("stars", 0),
            "logo_url": pkg.get("logo_image_id", ""),
            "repo_name": repo_info.get("name", ""),
            "repo_url": repo_info.get("url", ""),
            "official": repo_info.get("official", False),
        })
    return results


async def fetch_file(
    client: httpx.AsyncClient,
    org: str,
    repo: str,
    path: str,
    branch: str,
) -> str | None:
    """Fetch a file from GitHub API and return its decoded content."""
    url = f"https://api.github.com/repos/{org}/{repo}/contents/{path}"
    resp = await client.get(url, params={"ref": branch}, headers=_gh_headers())

    if resp.status_code == 404:
        return None
    if resp.status_code == 403:
        raise RuntimeError("GitHub API rate limit exceeded. Try again shortly.")
    resp.raise_for_status()

    data = resp.json()
    if data.get("encoding") == "base64":
        return base64.b64decode(data["content"]).decode("utf-8")
    return data.get("content", "")


async def fetch_chart(url: str) -> ChartInfo:
    """Fetch Chart.yaml and values.yaml from a GitHub Helm chart URL."""
    org, repo, branch, directory = parse_github_url(url)

    chart_path = f"{directory}/Chart.yaml" if directory != "." else "Chart.yaml"
    values_path = f"{directory}/values.yaml" if directory != "." else "values.yaml"

    async with httpx.AsyncClient(timeout=30.0) as client:
        chart_yaml = await fetch_file(client, org, repo, chart_path, branch)
        if chart_yaml is None:
            raise ValueError(
                f"No Chart.yaml found at {org}/{repo}/{chart_path}. "
                "Is this a Helm chart directory?"
            )

        values_yaml = await fetch_file(client, org, repo, values_path, branch)

    parsed = _parse_chart_yaml(chart_yaml)

    return ChartInfo(
        org=org,
        repo=repo,
        branch=branch,
        directory=directory,
        chart_yaml=chart_yaml,
        values_yaml=values_yaml,
        chart_name=parsed.get("name", ""),
        chart_version=parsed.get("version", ""),
        chart_description=parsed.get("description", ""),
        app_version=parsed.get("appVersion", ""),
        dependencies=parsed.get("dependencies", []),
    )


async def fetch_chart_from_artifacthub(repo_name: str, package_name: str) -> ChartInfo:
    """Fetch chart files from ArtifactHub by downloading the chart archive.

    Uses the ArtifactHub package API to get the content_url, then downloads
    and extracts Chart.yaml and values.yaml from the .tgz archive.
    """
    async with httpx.AsyncClient(timeout=30.0, follow_redirects=True) as client:
        # Get package details from ArtifactHub
        resp = await client.get(
            f"https://artifacthub.io/api/v1/packages/helm/{repo_name}/{package_name}",
        )
        if resp.status_code == 404:
            raise ValueError(f"Package {repo_name}/{package_name} not found on ArtifactHub.")
        resp.raise_for_status()
        pkg = resp.json()

        content_url = pkg.get("content_url", "")
        if not content_url:
            raise ValueError(f"No downloadable archive found for {repo_name}/{package_name}.")

        # Get the Helm registry URL for [helm_repo] block
        helm_repo_url = pkg.get("repository", {}).get("url", "")

        # Download the .tgz archive
        archive_resp = await client.get(content_url)
        archive_resp.raise_for_status()

    # Extract Chart.yaml and values.yaml from the archive
    chart_yaml = None
    values_yaml = None
    with tarfile.open(fileobj=io.BytesIO(archive_resp.content), mode="r:gz") as tar:
        for member in tar.getmembers():
            name = member.name
            # Files are typically at chartname/Chart.yaml or chartname/values.yaml
            basename = name.split("/", 1)[-1] if "/" in name else name
            if basename == "Chart.yaml" and chart_yaml is None:
                f = tar.extractfile(member)
                if f:
                    chart_yaml = f.read().decode("utf-8")
            elif basename == "values.yaml" and values_yaml is None:
                f = tar.extractfile(member)
                if f:
                    values_yaml = f.read().decode("utf-8")

    if chart_yaml is None:
        raise ValueError(f"No Chart.yaml found in archive for {repo_name}/{package_name}.")

    parsed = _parse_chart_yaml(chart_yaml)

    return ChartInfo(
        org=repo_name,
        repo=package_name,
        branch="",
        directory="",
        chart_yaml=chart_yaml,
        values_yaml=values_yaml,
        chart_name=parsed.get("name", package_name),
        chart_version=parsed.get("version", ""),
        chart_description=parsed.get("description", ""),
        app_version=parsed.get("appVersion", ""),
        dependencies=parsed.get("dependencies", []),
        helm_repo_url=helm_repo_url,
    )
